components/post_view/post_add_channel_member/post_add_channel_member.jsx

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {FormattedMessage} from 'react-intl';
+import {browserHistory} from 'react-router/es6';
+
+import store from 'stores/redux_store.jsx';
+
+import {Constants} from 'utils/constants.jsx';
+
+import AtMention from 'components/at_mention';
+
+const getState = store.getState;
+
+export default class PostAddChannelMember extends React.PureComponent {
+    static propTypes = {
+
+        /*
+        * Current team
+        */
+        team: PropTypes.object.isRequired,
+
+        /*
+        * Current channel
+        */
+        channel: PropTypes.object.isRequired,
+
+        /*
+        * post id of an ephemeral message
+        */
+        postId: PropTypes.string.isRequired,
+
+        /*
+        * user ids to add to channel
+        */
+        userIds: PropTypes.array.isRequired,
+
+        /*
+        * usernames to add to channel
+        */
+        usernames: PropTypes.array.isRequired,
+
+        /*
+        * User if has mention
+        */
+        hasMention: PropTypes.bool,
+
+        actions: PropTypes.shape({
+
+            /*
+            * Function to add members to channel
+            */
+            addChannelMember: PropTypes.func.isRequired,
+
+            /*
+            * Function to get post (ephemeral)
+            */
+            getPost: PropTypes.func.isRequired,
+
+            /*
+            * Function to remove post (ephemeral)
+            */
+            removePost: PropTypes.func.isRequired
+        }).isRequired
+    }
+
+    handleAddChannelMember = () => {
+        const {team, channel, postId, userIds} = this.props;
+        const post = this.props.actions.getPost(getState(), postId) || {};
+
+        if (post.channel_id === channel.id) {
+            userIds.forEach((userId) => {
+                this.props.actions.addChannelMember(channel.id, userId);
+            });
+
+            this.props.actions.removePost(post);
+        }
+
+        browserHistory.push(`/${team.name}/channels/${channel.name}`);
+    }
+
+    generateAtMentions(usernames = [], hasMention = false) {
+        if (usernames.length === 1) {
+            return (
+                <AtMention
+                    mentionName={usernames[0]}
+                    isRHS={false}
+                    hasMention={hasMention}
+                />
+            );
+        } else if (usernames.length > 1) {
+            function andSeparator(key) {
+                return (
+                    <FormattedMessage
+                        key={key}
+                        id={'post_body.check_for_out_of_channel_mentions.link.and'}
+                        defaultMessage={' and '}
+                    />
+                );
+            }
+
+            function commaSeparator(key) {
+                return <span key={key}>{', '}</span>;
+            }
```

Comment:
I don't think this isn't going to internationalize well. Not all languages have this sort of structure when listing names. Some have different rules about commas, and some don't use commas at all.

Reply:
Good to know that's the case for other language but I guess we're just using comma like in this one:
https://github.com/mattermost/mattermost-webapp/blob/master/components/post_view/reaction/reaction.jsx#L125